File: Javascript/10-Lasagna_Master/lasagna-master.js
/// <reference path="./global.d.ts" />
// @ts-check

export function cookingStatus(time) {
  if (time === 0) {
    return "Lasagna is done.";
  } else if (time === undefined) {
    return "You forgot to set the timer.";
  } else {
    return "Not done, please wait.";
  }
}

export function preparationTime(layers, time = 2) {
  return layers.length * time;
}

export function quantities(layers) {
  const noodles = layers.filter((layer) => layer === "noodles").length * 50;
  const sauce = layers.filter((layer) => layer === "sauce").length * 0.2;
  return { noodles: noodles, sauce: sauce };
}

export function addSecretIngredient(friendList, ownList) {
    ownList.push(friendList[friendList.length - 1]);
}

export function scaleRecipe(recipe, portions) {
    const scaledRecipe = {};
    Object.keys(recipe).forEach((ingredient) => {
        scaledRecipe[ingredient] = recipe[ingredient] * (portions / 2);
    });
    return scaledRecipe;
}

File: go/014-lasagna-master/lasagna_master.go
package lasagna

func PreparationTime(layers []string, timePerLayer int) int {
	if timePerLayer == 0 {
		return 2 * len(layers)
	}
	return timePerLayer * len(layers)
}

func Quantities(layers []string) (int, float64) {
	sauceLayers := float64(0)
	noodlesLayers := 0

	for _, v := range layers {
		if v == "sauce" {
			sauceLayers++
		} else if v == "noodles" {
			noodlesLayers++
		}
	}

	noodles := noodlesLayers * 50
	sauce := sauceLayers * 0.2

	return noodles, sauce
}

func AddSecretIngredient(ing1, ing2 []string) {
	ing2[len(ing2)-1] = ing1[len(ing1)-1]
}

func ScaleRecipe(quantities []float64, portions int) []float64 {
	quantitiesCopy := make([]float64, len(quantities))
	copy(quantitiesCopy, quantities)

	for i := range quantitiesCopy {
		quantitiesCopy[i] *= float64(portions) / 2
	}

	return quantitiesCopy
}
